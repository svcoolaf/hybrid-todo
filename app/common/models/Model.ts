export interface Model {
  create?: () => void;
  delete?: () => void;
  update?: () => void;
  id: number;
  toString?: (instance: Model) => string;
}
