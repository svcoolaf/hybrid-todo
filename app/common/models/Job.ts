import gql from "graphql-tag";
import { DefaultModel, Model } from "./Model";

export class Job extends DefaultModel implements Model {
  id: number;
  create = gql`
    mutation create($phone: String, $firstName: String, $lastName: String) {
      createClient(
        input: {
          client: { phone: $phone, firstName: $firstName, lastName: $lastName }
        }
      ) {
        client {
          id
          phone
          firstName
          lastName
        }
      }
    }
  `;
  delete: () => void;
  update: () => void;
}
