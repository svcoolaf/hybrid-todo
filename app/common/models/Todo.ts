import gql from "graphql-tag";
import { Model } from "./Model";

export class Todo implements Model {
  static create = gql`
    mutation create($content: String) {
      createJob(input: { job: { content: $content } }) {
        job {
          id
          content
        }
      }
    }
  `;
  static update = gql`
    mutation update($id: Int!, $content: String, $complete: Boolean) {
      updateJobById(
        input: { id: $id, jobPatch: { complete: $complete, content: $content } }
      ) {
        job {
          id
        }
      }
    }
  `;
  static delete = gql`
    mutation delete($id: Int!) {
      deleteJobById(input: { id: $id }) {
        job {
          id
        }
      }
    }
  `;
  static getAll = gql`
    {
      allJobs {
        nodes {
          id
          content
          complete
        }
      }
    }
  `;
  id: number;
  content: string;
  complete: boolean;
}
