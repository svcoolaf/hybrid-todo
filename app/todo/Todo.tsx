import React from "react";

const Todo = () => <>Todo</>;

export default Todo;
