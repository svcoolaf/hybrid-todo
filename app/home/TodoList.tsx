import { useMutation, useQuery } from "@apollo/react-hooks";
import React, { useEffect, useState } from "react";
import styled from "styled-components";
import theme from "../common/lib/theme";
import { Model } from "../common/models/Model";
import { Todo } from "../common/models/Todo";
import Check from "./Check";
import FloatingForm from "./FloatingForm";
import Hero from "./Hero";
import Icon from "./Icon";

const DeleteButton = styled.button`
  display: block;
  background: transparent;
  border: none;
`;
const Left = styled.div`
  display: flex;
  align-items: center;
`;
const TodoItem = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  cursor: pointer;
  padding: ${theme.sizes.double}px;
  border-bottom: 1px ${theme.colors.light} solid;

  .delete-button {
    cursor: pointer;
    opacity: 0;
    transition: opacity ${theme.animation.time} ease-out;
  }
  &:hover {
    .delete-button {
      opacity: 1;
    }
  }
`;

const Content = styled.span`
  padding: ${theme.sizes.double}px;
  display: block;
`;

const TodoList = () => {
  const [createJob] = useMutation<any, Partial<Todo>>(Todo.create);
  const [updateTodo] = useMutation<any, Partial<Todo>>(Todo.update);
  const [deleteJob] = useMutation<any, Model>(Todo.delete);
  const [update, setUpdate] = useState(0);
  const [content, setContent] = useState("");
  const triggerRefetch = () => setUpdate(update + 1);
  const { loading, data, error, refetch } = useQuery(Todo.getAll);
  const todos = (data.allJobs || {}).nodes;

  useEffect(() => {
    refetch();
  }, [update]);

  if (loading) {
    return <p>Loading...</p>;
  }
  if (error) {
    return <p>Error :(</p>;
  }
  const getAddTodoHandler = job => e => {
    e.preventDefault();
    if (e.key === "d") {
      return getDeleteHandler(job)();
    }
    updateTodo({
      variables: { id: job.id, complete: !job.complete }
    }).then(triggerRefetch);
  };
  const getDeleteHandler = job => () => {
    deleteJob({
      variables: job
    }).then(triggerRefetch);
  };

  return (
    <>
      <Hero />
      {todos.map((job: Todo) => (
        <TodoItem
          key={job.id}
          onClick={getAddTodoHandler(job)}
          onKeyPress={getAddTodoHandler(job)}
          tabIndex={0}
        >
          <Left>
            <Check onClick={getAddTodoHandler(job)} checked={job.complete} />
            <Content>{job.content}</Content>
          </Left>
          <DeleteButton
            className="delete-button"
            onClick={getDeleteHandler(job)}
            tabIndex={0}
          >
            <Icon name="cross" />
          </DeleteButton>
        </TodoItem>
      ))}
      <FloatingForm
        onFormSubmit={e => {
          e.preventDefault();
          if (!content) {
            return;
          }
          createJob({
            variables: { content }
          }).then(() => {
            setContent("");
            triggerRefetch();
          });
        }}
        content={content}
        setContent={setContent}
      />
    </>
  );
};

export default TodoList;
