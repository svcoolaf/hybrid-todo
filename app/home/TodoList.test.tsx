import { MockedProvider, wait } from "@apollo/react-testing";
import { noop } from "lodash";
import React from "react";
import renderer from "react-test-renderer";
import { Todo } from "../common/models/Todo";
import TodoList from "./TodoList";

const mocks = [
  {
    request: {
      query: Todo.getAll
    },
    result: {
      data: {
        allJobs: {
          nodes: [
            {
              id: 13,
              content: "1",
              complete: false
            },
            {
              id: 19,
              content: "2",
              complete: false
            },
            {
              id: 23,
              content: "3",
              complete: false
            }
          ]
        }
      }
    }
  }
];

describe("TodoList", () => {
  it("renders", async () => {
    const tree = renderer.create(
      <MockedProvider mocks={mocks} addTypename={false}>
        <TodoList triggerFetch={noop} fetchHash={0} />
      </MockedProvider>
    );
    await wait(0);

    expect(tree.toJSON()).toMatchSnapshot();
  });
});
