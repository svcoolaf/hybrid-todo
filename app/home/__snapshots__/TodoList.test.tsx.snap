// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList renders 1`] = `
Array [
  <div
    className="TodoList__TodoItem-m2x2f9-2 ksHrKr"
    onClick={[Function]}
    onKeyPress={[Function]}
    tabIndex={0}
  >
    <div
      className="TodoList__Left-m2x2f9-1 bJSXil"
    >
      <button
        checked={false}
        className="Check__RoundButton-sc-1sbkphr-0 yCVmH"
        onClick={[Function]}
        tabIndex={0}
      >
        <div
          className="Check__Placeholder-sc-1sbkphr-1 jCOvgI"
        />
      </button>
      <span
        className="TodoList__Content-m2x2f9-3 fHTCag"
      >
        1
      </span>
    </div>
    <button
      className="delete-button TodoList__DeleteButton-m2x2f9-0 hyDOLv"
      onClick={[Function]}
      tabIndex={0}
    >
      <img
        alt="cross"
        height={24}
        src="/static/icons/cross.svg"
        width={24}
      />
    </button>
  </div>,
  <div
    className="TodoList__TodoItem-m2x2f9-2 ksHrKr"
    onClick={[Function]}
    onKeyPress={[Function]}
    tabIndex={0}
  >
    <div
      className="TodoList__Left-m2x2f9-1 bJSXil"
    >
      <button
        checked={false}
        className="Check__RoundButton-sc-1sbkphr-0 yCVmH"
        onClick={[Function]}
        tabIndex={0}
      >
        <div
          className="Check__Placeholder-sc-1sbkphr-1 jCOvgI"
        />
      </button>
      <span
        className="TodoList__Content-m2x2f9-3 fHTCag"
      >
        2
      </span>
    </div>
    <button
      className="delete-button TodoList__DeleteButton-m2x2f9-0 hyDOLv"
      onClick={[Function]}
      tabIndex={0}
    >
      <img
        alt="cross"
        height={24}
        src="/static/icons/cross.svg"
        width={24}
      />
    </button>
  </div>,
  <div
    className="TodoList__TodoItem-m2x2f9-2 ksHrKr"
    onClick={[Function]}
    onKeyPress={[Function]}
    tabIndex={0}
  >
    <div
      className="TodoList__Left-m2x2f9-1 bJSXil"
    >
      <button
        checked={false}
        className="Check__RoundButton-sc-1sbkphr-0 yCVmH"
        onClick={[Function]}
        tabIndex={0}
      >
        <div
          className="Check__Placeholder-sc-1sbkphr-1 jCOvgI"
        />
      </button>
      <span
        className="TodoList__Content-m2x2f9-3 fHTCag"
      >
        3
      </span>
    </div>
    <button
      className="delete-button TodoList__DeleteButton-m2x2f9-0 hyDOLv"
      onClick={[Function]}
      tabIndex={0}
    >
      <img
        alt="cross"
        height={24}
        src="/static/icons/cross.svg"
        width={24}
      />
    </button>
  </div>,
]
`;
